src/java/org/apache/cassandra/index/sai/disk/v2/sortedterms/SortedTermsReader.java

```diff
@@ -0,0 +1,303 @@
+     * <p>
+     * The cursor is to be used in a single thread.
+     * The cursor is valid as long this object hasn't been closed.
+     * You must close the cursor when you no longer need it.
+     */
+    public @Nonnull Cursor openCursor() throws IOException
+    {
+        return new Cursor(termsData);
+    }
+
+    /**
+     * Allows reading the terms from the terms file.
+     * Can quickly seek to a random term by <code>pointId</code>.
+     * <p>
+     * This object is stateful and not thread safe.
+     * It maintains a position to the current term as well as a buffer that can hold one term.
+     */
+    @NotThreadSafe
+    public class Cursor implements AutoCloseable
+    {
+        private final IndexInput termsData;
+
+        // The term the cursor currently points to. Initially empty.
+        private final BytesRef currentTerm;
+
+        // The point id the cursor currently points to. -1 means before the first item.
+        private long pointId = -1;
+
+        Cursor(IndexInput termsData) throws IOException
+        {
+            this.termsData = termsData.clone();  // clone needed to get our own file pointer independently of other Cursors
+            this.termsData.seek(termsDataStartFP);
+            this.currentTerm = new BytesRef(meta.maxTermLength);
+        }
+
+        /**
+         * Returns the number of terms
+         */
+        public long count()
+        {
+            return SortedTermsReader.this.count();
+        }
+
+        /**
+         * Returns the current position of the cursor.
+         * Initially, before the first call to {@link Cursor#advance}, the cursor is positioned at -1.
+         * After reading all the items, the cursor is positioned at index one
+         * greater than the position of the last item.
+         */
+        public long pointId()
+        {
+            return pointId;
+        }
+
+        /**
+         * Returns the current term data as <code>ByteComparable</code> referencing the internal term buffer.
+         * The term data stored behind that reference is valid only until the next call to
+         * {@link Cursor#advance} or {@link Cursor#seekToPointId(long)}.
+         */
+        public @Nonnull ByteComparable term()
+        {
+            return ByteComparable.fixedLength(currentTerm.bytes, currentTerm.offset, currentTerm.length);
+        }
+
+        /**
+         * Moves the cursor to the next term and reads it into the current term buffer.
+         * <p>
+         * If there are no more available terms, clears the term buffer and the cursor's position will point to the
+         * one behind the last item.
+         * <p>
+         * This method has constant time complexity.
+         *
+         * @return true if the cursor was moved successfully, false if end of file was reached
+         * @throws IOException if a read from the terms file fails
+         */
+        public boolean advance() throws IOException
```

Comment:
The only place the `boolean` return is used is in tests. Is that correct? If so it might be worth mentioning or removing the return and changing the tests to use the same condition testing as the live code.

Reply:
Returning boolean here enables easy iteration with a while loop:

```java
while (cursor.advance()) {
    // do sth with cursor.term()
}
```

Sure, we don't use it now in other places than tests, but it makes the abstraction really much more useful and at the near-zero cost.

I'll add an assert on return value in `seekToPointId`.